rust/kernel/sync/locked_by.rs

```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! A wrapper for data protected by a lock that does not wrap it.
+
+use super::{Guard, Lock};
+use core::{cell::UnsafeCell, ops::Deref, ptr};
+
+/// Allows access to some data to be serialised by a lock that does not wrap it.
+///
+/// In most cases, data protected by a lock is wrapped by the appropriate lock type. [`LockedBy`]
+/// is meant for cases when this is not possible. For example, if a container has a lock and some
+/// data in the contained elements needs to be protected by the same lock.
```

Comment:
A few more words or a particular example would probably be helpful: I assume you are referring to a data structure where only a subset of the members of the contained elements need to be locked by the same lock? i.e. where a subset of the members need not be locked to be accessed.

In that case, it may also raise questions: e.g. one could perhaps rearrange the data structure to keep all locked members on one side and the non-locked parts in another (e.g. struct of arrays vs. arrays of structs) so that a normal `Mutex<T>` could be used around the data that needs locking. Of course, that doesn't always work if one really benefits from a particular memory arrangement, but if it does not matter (e.g. for locality), then it is best to use the wrapped data lock instead.

Reply:
i.e. preemptively answering this kind of reasoning in the docs is probably useful.

Reply:
Ah, I see -- yeah, it definitely does!